### PR TITLE
Pretrained contraction metric

### PR DESCRIPTION
**Pull request by dtch1997**
Goal: investigate whether it is possible to disentangle the learning of the contraction metric from the learning of the tracking controller. Learning will proceed in a 2-stage process:
(1) Pre-train $M$ on a dataset of trajectory bundles, with each bundle consisting of pairwise converging trajectories, and the state dynamics can be simply estimated from the trajectory sequence. 
(2) Train $u$ as before, but with a frozen $M$.

### CODE REVIEW

**Comment by dtch1997** [comment]
It would be interesting to test the performance difference between pre-trained and frozen M, as well as pre-trained and learnable (or fine-tunable) M.

**Comment by dtch1997** [comment]
Just realized there's a bug in the calculation of `xdot` in `TwoStageTrainer`. I've been using `sim_dt` but it should actually be `controller_dt` with the current implementation. 

Actually it would probably better to toss that out and replace it with an implementation based on `sim_dt`.

**Comment by dtch1997** [comment]
I would like to obtain a comparison plot of the learned controller with pre-training vs without, therefore I will turn off global step increment in the pre-training stage

**Comment by dtch1997** [comment]
We need to refactor the logging to clarify certain quantities. (This list is subject to updates)

During pretraining, log:
- The PD loss
- The M_loss using the simulated `xdot` 

During policy training, log:
- PD loss
- M_loss from simulated `x_dot` 
- M_loss from policy `x_dot` --> this measures how well the policy is learning to track the reference

**Comment by dtch1997** [comment]
Conclusion from the pretrain-and-finetune run: https://wandb.ai/dtch1997/SkillVerification/runs/gxgonu8h?workspace=user-dtch1997
Pre-training of contraction metric seemed to work well. Contraction metric loss had decreased to about 0.3 at start of training. Policy learning also seemed to go well for a while, but at the end the M_loss (due to policy) somehow began to rise again. There seems to be some sort of 'overfitting' to the frozen contraction metric taking place. 

An interesting artefact during policy learning: On one simulated attempt, the policy inexplicably veered off-course, accompanied by corresponding increase in the metric. 

metric_deviation
phase_plane_deviation

**Comment by dtch1997** [comment]
The next step is to try and visualize the regions of the state space where the learned metric is valid. Let's try to formalize this a little bit. Assume that we have a reference-conditioned autonomous dynamics function $\dot x = g(x, x^*)$, which is equivalent to the case of a fixed controller and reference trajectory. 

The state space consists of $x = (p_x, p_y, \theta)$. $M(x)$ defines a bilinear map. The metric proves that $x$ is converging to $x^\*$ if for all $x_p$ in a small ball $B_\epsilon(x)$ we have that $\frac{d}{dt} d(x_p, x^\*) < 0$, where $d(a, b) = (a-b)^TM(a)(a-b)$. 

To evaluate this empirically requires knowing $\dot x_p$, i.e. querying our controller at the perturbed spot and also being able to simulate it. 

An alternative approach may be to use a Lipschitz-like condition to analytically validate the metric. Intuitively, if $d/dx M(x)$ is well-behaved enough and $\frac{d}{dt} d(x, x^\*) < 0$ then we should be able to say directly that $\frac{d}{dt} d(x_p, x^\*) < 0$ for all $x_p$ "close enough". TODO: make this argument concrete. 

Assuming we are able to do this, then a suitable evaluation metric could be proportion of the state space where learned contraction metric is valid.

**Comment by dtch1997** [comment]
I'm not really sure whether this works, but it doesn't break anything with the original workflow, so I'm just going to merge it in as an experimental feature and document the remaining issues
